### DISCUSSION TITLE
Creating a new Image<TPixel> by Wrapping IMemoryOwner<byte>

### DISCUSSION THREAD

**Discussion by ptasev** [SixLabors/ImageSharp#1580] [Ideas] [answered]
What are your thoughts on adding a WrapMemory overload that takes in `IMemoryOwner<byte>`?  One scenario I have come across that this would be helpful is when having to manually decompress some pixel data in byte form before passing over ownership to the image.  I'm using `ArrayPoolBufferWriter<byte>` from the MS HighPerf toolkit to minimize copying that would typically be needed in this situation when using MemoryStream.

```csharp
// showing usings of third-party libs
using ICSharpCode.SharpZipLib.Zip.Compression.Streams;
using Microsoft.Toolkit.HighPerformance;
using Microsoft.Toolkit.HighPerformance.Buffers;

private static IMemoryOwner<byte> ZlibDecompress(ReadOnlyMemory<byte> data)
{
    using (var ss = data.AsStream())
    using (var iis = new InflaterInputStream(ss))
    {
        var bufferWriter = new ArrayPoolBufferWriter<byte>(data.Length);

        // Inflate
        iis.CopyTo(bufferWriter.AsStream());

        return bufferWriter;
    }
}

// showing usage code
var compressedData = ...;
IMemoryOwner<byte> decompressedMemoryOwner = ZlibDecompress(compressedData);
// This does not work, expects IMemoryOwner<TPixel>
var image = Image.WrapMemory<TPixel>(decompressedMemoryOwner, width, height);
```

**Comment by antonfirsov** [comment, marked as answer]
This would be essentially a convenience extension accompanying an existing `IMemoryOwner<TPixel>` method:

https://github.com/SixLabors/ImageSharp/blob/08b0320538636d369b50fcfce6e3fb2b8404a8d9/src/ImageSharp/Image.WrapMemory.cs#L140-L147

Since `Memory<byte>` overloads already exist, it makes sense to also have `IMemoryOwner<byte>` overloads for API symmetry. The easiest implementation would be to create a `WrapperMemoryOwner<TPixel>` that adapts `IMemoryOwner<byte>`. Interested in PR-ing it?

/cc @Sergio0694

**Reply by Sergio0694** [reply]
The first issue I see with someone using the `HighPerformance` package, this warms my heart 😍

@ptasev Let us know if you'd like to make a PR, otherwise I could see if I have some time to add this at some point!

**Reply by ptasev** [reply]
Lacking the motivation to do more programming after work at the moment so it's all yours if you'd like.

Also, thanks for all your hard work. That package is super useful.
